orttraining/orttraining/training_ops/cuda/communication/send.cc

```diff
@@ -3,69 +3,117 @@
+  std::vector<size_t> tensor_sizes_in_bytes;
+
+  // Compute tensor shapes and sizes
+  size_t sum = 0;
+  for (int i = 0; i < tensor_num; ++i) {
+    const Tensor* x_tensor = ctx->Input<Tensor>(i + 2);
+    sum += x_tensor->Shape().NumDimensions();
+    prefix_tensor_shape_sizes.push_back(sum);
+    aggregated_tensor_shapes.insert(aggregated_tensor_shapes.end(),
+                                    x_tensor->Shape().GetDims().begin(),
+                                    x_tensor->Shape().GetDims().end());
+
+    // handle alignment requirement
+    aggregated_aligned_tensor_bytes = (aggregated_aligned_tensor_bytes + alignment - 1) / alignment * alignment;
+    tensor_offsets_in_bytes.push_back(aggregated_aligned_tensor_bytes);
+    aggregated_aligned_tensor_bytes += x_tensor->SizeInBytes();
+    tensor_sizes_in_bytes.push_back(x_tensor->SizeInBytes());
+  }
+
+  // Start communication
   int world_rank;
   MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
-  
-  ORT_ENFORCE(world_rank == src_, "Sending data from rank ", world_rank, " but the expected data is on rank ", src_, ".");
-
-  // TODO: we need to avoid copying data to host after enabling CUDA-aware MPI in the build.
-  IAllocatorUniquePtr<T> buffer = AllocateBufferOnCPUPinned<T>(tensor_size);
-  ORT_ENFORCE(cudaMemcpy(
-    buffer.get(), x_tensor->Data<T>(), tensor_size * sizeof(T), cudaMemcpyDeviceToHost) == cudaSuccess);
-
-  // Prepare communication information.
-  std::vector<int64_t> tensor_shape(x_tensor->Shape().GetDims());
-  CommInfo_t info_shape_size{&tensor_shape_size, sizeof(size_t), static_cast<int>(dst_), static_cast<int>(tag_)};
-  CommInfo_t info_shape{tensor_shape.data(), tensor_shape_size * sizeof(int64_t), static_cast<int>(dst_), static_cast<int>(tag_)};
-  CommInfo_t info_data{buffer.get(), tensor_size * sizeof(T), static_cast<int>(dst_), static_cast<int>(tag_)};
-
-  // Enqueue communication functions to GPU stream.
+  ORT_ENFORCE(world_rank != dst, "Sending data to rank ", dst, " on the rank ", world_rank, ".");
+
+  IAllocatorUniquePtr<char> buffer = AllocateBufferOnCPUPinned<char>(
+      static_cast<size_t>(aggregated_aligned_tensor_bytes));
+
+  // Keep the sync copy in the previous design
+  // TODO they can be moved to async call after global stream becoming accessible
+  for (int i = 0; i < tensor_num; ++i) {
+    const Tensor* x_tensor = ctx->Input<Tensor>(i + 2);
+    ORT_ENFORCE(cudaMemcpy(buffer.get() + tensor_offsets_in_bytes[i], x_tensor->Data<void>(),
+                           tensor_sizes_in_bytes[i], cudaMemcpyDeviceToHost) == cudaSuccess);
+  }
+
+  // Prepare MPI communication info
+  CommInfo_t info_shape_sizes{prefix_tensor_shape_sizes.data(),
+                              tensor_num * static_cast<int>(sizeof(size_t)),
+                              dst,
+                              static_cast<int>(tag_)};
+  CommInfo_t info_aggregated_size{&aggregated_aligned_tensor_bytes,
+                                  static_cast<int>(sizeof(size_t)),
```

Comment:
`int` here is a bit dangerous.

Reply:
static_cast<int>(sizeof(size_t)) is dangerous??
why?
It most likely will be optimized to a literal. 
i did static_cast, simply because of warning.
struct Initializer should match type.
Its type is changed to int from size_t in this PR, since MPI spec use int.

Reply:
I will keep this unchanged for now.

Reply:
I don't think it's safe.
```
aggregated_aligned_tensor_bytes += x_tensor->SizeInBytes();
```
We have 20B model now and it will only be larger.

Reply:
If your argument is that Send/Recv is not designed for large tensors communication, please throw if overflow happens.

Reply:
If that is the case, it should not be this line, right?
The problem is not overflow.
The problem is MPI's size has a upper-bound, which is int's upper-bound.
MPI's size only support int based on the spec.
That means the entire MPI should be serialized.

20B model should be fine, from my view.
Pipeline is used to partition model to solve memory issue.
The GPU memory probably hit the bound earlier than MPI's bound.

Reply:
Previously we don't care because we can only send/recv one tensor.

Reply:
Well, I can do a safety check for aggregated_aligned_tensor_bytes <=int_max.
I don't think we want to do serialization now.